statsmodels/discrete/discrete_model.py

```diff
@@ -1173,6 +1174,155 @@ def hessian(self, params):
         L = np.exp(np.dot(X,params) + exposure + offset)
         return -np.dot(L*X.T, X)
 
+class GeneralizedPoisson(CountModel):
+    __doc__ = """
+    Generalized Poisson model for count data
+
+    %(params)s
+    %(extra_params)s
+
+    Attributes
+    -----------
+    endog : array
+        A reference to the endogenous response variable
+    exog : array
+        A reference to the exogenous design.
+    """ % {'params' : base._model_params_doc,
+           'extra_params' :
+           """offset : array_like
+        Offset is added to the linear prediction with coefficient equal to 1.
+    exposure : array_like
+        Log(exposure) is added to the linear prediction with coefficient
+        equal to 1.
+
+    """ + base._missing_param_doc}
+
+    def __init__(self, endog, exog, p = 1, offset=None,
+                       exposure=None, missing='none', **kwargs):
+        super(GeneralizedPoisson, self).__init__(endog, exog, offset=offset,
+                                               exposure=exposure,
+                                               missing=missing, **kwargs)
+        self.p = p - 1
+
+    def loglike(self, params):
+        """
+        Loglikelihood of Generalized Poisson model
+
+        Parameters
+        ----------
+        params : array-like
+            The parameters of the model.
+
+        Returns
+        -------
+        loglike : float
+            The log-likelihood function of the model evaluated at `params`.
+            See notes.
+
+        Notes
+        --------
+        """
+        return np.sum(self.loglikeobs(params))
+
+    def loglikeobs(self, params):
+        """
+        Loglikelihood for observations of Generalized Poisson model
+
+        Parameters
+        ----------
+        params : array-like
+            The parameters of the model.
+
+        Returns
+        -------
+        loglike : ndarray (nobs,)
+            The log likelihood for each observation of the model evaluated
+            at `params`. See Notes
+
+        Notes
+        --------
+        """
+        alpha = params[-1]
+        p = self.p
+        endog = self.endog
+        mu = self.predict(params[:-1])
+        return (np.log(mu) + (endog - 1) * np.log(mu + alpha * np.power(mu, p) * endog) - 
+                np.log(1 + alpha * np.power(mu, p)) - gammaln(endog + 1) - 
+                (mu + alpha * np.power(mu, p) * endog) / (1 + alpha * np.power(mu, p)))
```

Comment:
@jbrockmendel 
gammaln is the best function for this kind of computation.
It has one of the best precision and avoids overflow (and is heavily used in scipy.stats.distributions for this reason).
Also, we don't want to restrict to integers. (for quasi-maximum likelihood, endog could be any non-negative number.

About caching intermediate computation that are invariant in the optimization. This is a bit tricky in general and I think we should skip it in the first version of this PR/GSOC.
https://github.com/statsmodels/statsmodels/pull/3692#issuecomment-304060856
(If it turns out that it looks relatively easy for a model, then we could try it out later.)